### PR TITLE
Issue 2895: Create Wire protocol changes for TableStore apis.

### PR DESCRIPTION
**Pull request by shrids**
**Change log description**  
This PR has the WireProtocol  changes and the corresponding implementation on PravegaRequestProcessor for the table store apis apart from the keyIterator and entryIterator. 

**Purpose of the change**  
Issue #2895 

**What the code does**  
WireProtocol changes and the corresponding request processor implementation for the below table store apis.

```
CompletableFuture<Void> createSegment(String segmentName, Duration timeout);
CompletableFuture<Void> deleteSegment(String segmentName, boolean mustBeEmpty, Duration timeout);
CompletableFuture<Void> merge(String targetSegmentName, String sourceSegmentName, Duration timeout);
CompletableFuture<Void> seal(String segmentName, Duration timeout);
CompletableFuture<List<Long>> put(String segmentName, List<TableEntry> entries, Duration timeout);
CompletableFuture<Void> remove(String segmentName, Collection<TableKey> keys, Duration timeout);
CompletableFuture<List<TableEntry>> get(String segmentName, List<ArrayView> keys, Duration timeout); 
```

**How to verify it**  
All the existing and newly added tests should continue to pass.

### CODE REVIEW

**Comment by shrids** [comment]
[WireProtocol changes for the tables API work.](https://trello.com/c/c527QYL6/75-wireprotocol-changes-for-the-tables-api-work)

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java [hunk 1]
```diff
@@ -552,6 +571,144 @@ public void updateSegmentPolicy(UpdateSegmentPolicy updateSegmentPolicy) {
...
+    @Override
+    public void mergeTableSegments(final MergeTableSegments mergeTableSegments) {
+        final String operation = "mergeTableSegments";
+
+        if (!verifyToken(mergeTableSegments.getSource(), mergeTableSegments.getRequestId(), mergeTableSegments.getDelegationToken(), operation)) {
+            return;
+        }
+
+        log.info(mergeTableSegments.getRequestId(), "Merging table segments {}.", mergeTableSegments);
+        tableStore.merge(mergeTableSegments.getTarget(), mergeTableSegments.getSource(), TIMEOUT)
```
Comment (tkaitchuck) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: Do we also want to support conditional merge?

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java [hunk 2]
```diff
@@ -281,6 +284,43 @@ public boolean isFailure() {
         }
     }
 
+    @Data
+    public static final class NotEmptyTableSegment implements Reply, WireCommand {
```
Comment (tkaitchuck) [inline shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: I think this should be called "TableSegmentNotEmpty"

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java [hunk 3]
```diff
@@ -1146,7 +1245,7 @@ public void process(RequestProcessor cp) {
         public void writeFields(DataOutput out) throws IOException {
             out.writeLong(requestId);
             out.writeUTF(segment);
-            out.writeUTF(delegationToken == null ? "" : delegationToken);
+            out.writeUTF(delegationToken);
```
Comment (tkaitchuck) [inline shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: Why this change? The Javadoc says this will throw an NPE if it's null.

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java [hunk 4]
```diff
@@ -1262,6 +1388,36 @@ public static WireCommand readFrom(DataInput in, int length) throws IOException
...
+        public void process(RequestProcessor cp) {
+            cp.deleteTableSegment(this);
+        }
+
+        @Override
+        public void writeFields(DataOutput out) throws IOException {
+            out.writeLong(requestId);
+            out.writeUTF(segment);
+            out.writeBoolean(mustBeEmpty);
+            out.writeUTF(delegationToken);
```
Comment (tkaitchuck) [inline shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: null check?

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java [hunk 5]
```diff
@@ -990,6 +1030,35 @@ public static WireCommand readFrom(DataInput in, int length) throws IOException
...
+        @Override
+        public void process(RequestProcessor cp) {
+            cp.createTableSegment(this);
+        }
+
+        @Override
+        public void writeFields(DataOutput out) throws IOException {
+            out.writeLong(requestId);
+            out.writeUTF(segment);
+            out.writeUTF(delegationToken);
```
Comment (tkaitchuck) [inline shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: Null check?

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java [hunk 6]
```diff
@@ -1103,6 +1172,36 @@ public static WireCommand readFrom(DataInput in, int length) throws IOException
...
+        public void process(RequestProcessor cp) {
+            cp.mergeTableSegments(this);
+        }
+
+        @Override
+        public void writeFields(DataOutput out) throws IOException {
+            out.writeLong(requestId);
+            out.writeUTF(target);
+            out.writeUTF(source);
+            out.writeUTF(delegationToken);
```
Comment (tkaitchuck) [inline shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: null check?

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java [hunk 7]
```diff
@@ -1349,4 +1505,327 @@ public static WireCommand readFrom(ByteBufInputStream in, int length) throws IOE
...
+        @Override
+        public void process(RequestProcessor cp) {
+            cp.updateTableEntries(this);
+        }
+
+        @Override
+        public void writeFields(DataOutput out) throws IOException {
+            out.writeLong(requestId);
+            out.writeUTF(segment);
+            out.writeUTF(delegationToken);
```
Comment (tkaitchuck) [inline shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: null check?

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java [hunk 8]
```diff
@@ -1349,4 +1505,327 @@ public static WireCommand readFrom(ByteBufInputStream in, int length) throws IOE
...
+        @Override
+        public void process(RequestProcessor cp) {
+            cp.readTable(this);
+        }
+
+        @Override
+        public void writeFields(DataOutput out) throws IOException {
+            out.writeLong(requestId);
+            out.writeUTF(segment);
+            out.writeUTF(delegationToken);
```
Comment (tkaitchuck) [inline shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: null check?

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java [hunk 9]
```diff
@@ -1349,4 +1505,327 @@ public static WireCommand readFrom(ByteBufInputStream in, int length) throws IOE
...
+            long requestId = in.readLong();
+            String segment = in.readUTF();
+            TableEntries entries = (TableEntries) TableEntries.readFrom(in, in.available());
+            return new TableRead(requestId, segment, entries);
+        }
+    }
+
+    @Data
+    public static final class TableEntries implements WireCommand {
+        final WireCommandType type = WireCommandType.TABLE_ENTRIES;
```
Comment (tkaitchuck) [inline shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: It looks to me like TableEntries is just a convenience for embedding into other commands. In which case it doesn't need it's own WireCommandType because it's not actually used.

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java [hunk 10]
```diff
@@ -1349,4 +1505,327 @@ public static WireCommand readFrom(ByteBufInputStream in, int length) throws IOE
...
+            }
+
+            return new TableEntries(entries);
+        }
+    }
+
+    @Data
+    public static final class TableKey implements WireCommand {
+        public final static TableKey EMPTY = new TableKey(ByteBuffer.wrap(new byte[0]), Long.MIN_VALUE);
+        final WireCommandType type = WireCommandType.TABLE_KEY;
```
Comment (tkaitchuck) [inline shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: Same here

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java [hunk 11]
```diff
@@ -1349,4 +1505,327 @@ public static WireCommand readFrom(ByteBufInputStream in, int length) throws IOE
...
+            in.readFully(msg);
+            long keyVersion = in.readLong();
+            return new TableKey(ByteBuffer.wrap(msg), keyVersion);
+        }
+    }
+
+    @Data
+    public static final class TableValue implements WireCommand {
+        public final static TableValue EMPTY = new TableValue(ByteBuffer.wrap(new byte[0]));
+        final WireCommandType type = WireCommandType.TABLE_VALUE;
```
Comment (tkaitchuck) [inline shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: Same here

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java [hunk 12]
```diff
@@ -1349,4 +1505,327 @@ public static WireCommand readFrom(ByteBufInputStream in, int length) throws IOE
...
+    @Data
+    public static final class TableKey implements WireCommand {
+        public final static TableKey EMPTY = new TableKey(ByteBuffer.wrap(new byte[0]), Long.MIN_VALUE);
+        final WireCommandType type = WireCommandType.TABLE_KEY;
+        final ByteBuffer data;
+        final long keyVersion;
+
+        @Override
+        public void writeFields(DataOutput out) throws IOException {
+            out.writeInt(type.getCode());
```
Comment (tkaitchuck) [inline shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: This shouldn't be needed.

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java [hunk 13]
```diff
@@ -1349,4 +1505,327 @@ public static WireCommand readFrom(ByteBufInputStream in, int length) throws IOE
...
+
+    @Data
+    public static final class TableEntries implements WireCommand {
+        final WireCommandType type = WireCommandType.TABLE_ENTRIES;
+        final List<Map.Entry<TableKey, TableValue>> entries;
+
+        @Override
+        public void writeFields(DataOutput out) throws IOException {
+            out.writeInt(entries.size());
+            for (Map.Entry<TableKey, TableValue> ent :entries) {
```
Comment (tkaitchuck) [inline shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: The way these table keys and values are being written makes it impossible to add more fields to table key and value in the future because one will end and any unread bytes will be assumed to be part of the next item in entries.

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java [hunk 14]
```diff
@@ -1349,4 +1505,327 @@ public static WireCommand readFrom(ByteBufInputStream in, int length) throws IOE
...
+            cp.readTable(this);
+        }
+
+        @Override
+        public void writeFields(DataOutput out) throws IOException {
+            out.writeLong(requestId);
+            out.writeUTF(segment);
+            out.writeUTF(delegationToken);
+            out.writeInt(keys.size());
+            for (TableKey key : keys) {
```
Comment (tkaitchuck) [inline shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: See my above comment about versioning.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java (hunk 1 above)
Comment (shrids) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: tableStore.merge() is currently not implemented and will throw a `UnsupportedOperationException`

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java (hunk 2 above)
Comment (shrids) [reply shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: done.

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java (hunk 3 above)
Comment (shrids) [reply shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: removed this unnecessary change.

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java (hunk 4 above)
Comment (shrids) [reply shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: added it for all the new commands.

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java (hunk 9 above)
Comment (shrids) [reply shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: yes, it is a convenience class  and should not need its own WireCommandType. Removed it.

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java (hunk 12 above)
Comment (shrids) [reply shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: removed it.

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java (hunk 13 above)
Comment (shrids) [reply shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: Added payload size field to handle future updates to TableKey and TableValue.

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java (hunk 14 above)
Comment (shrids) [reply shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: Added payload size to handle future updates to TableKey.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java [hunk 15]
```diff
@@ -552,6 +571,144 @@ public void updateSegmentPolicy(UpdateSegmentPolicy updateSegmentPolicy) {
...
+                  .thenAccept(values -> connection.send(new WireCommands.TableRead(readTable.getRequestId(), segment, getTableEntries(values))))
+                  .exceptionally(e -> handleException(readTable.getRequestId(), segment, operation, e));
+    }
+
+    private WireCommands.TableEntries getTableEntries(final List<TableEntry> updateData) {
+
+        final List<Map.Entry<WireCommands.TableKey, WireCommands.TableValue>> entries = updateData.stream().map(te -> {
+            if (te == null) {
+                // no entry for the corresponding key.
+                return new AbstractMap.SimpleImmutableEntry<>(WireCommands.TableKey.EMPTY, WireCommands.TableValue.EMPTY);
```
Comment (shrids) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: TODO: the key  should correspond to the key passed in the request.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java [hunk 16]
```diff
@@ -552,6 +571,144 @@ public void updateSegmentPolicy(UpdateSegmentPolicy updateSegmentPolicy) {
...
+        String segment = updateTableEntries.getSegment();
+        final String operation = "updateTableEntries";
+
+        if (!verifyToken(segment, updateTableEntries.getRequestId(), updateTableEntries.getDelegationToken(), operation)) {
+            return;
+        }
+
+        log.info(updateTableEntries.getRequestId(), "Updating table segment {}.", updateTableEntries);
+        List<TableEntry> entries = updateTableEntries.getTableEntries().getEntries().stream()
+                                                     .map(e -> versioned(new ByteArraySegment(e.getKey().getData()),
```
Comment (andreipaduroiu) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: Not all of these are `versioned`. We need to make the following transformations:
1. If `Client.TableKey.getVersion() == null` -> use `TableEntry.unversioned`
2. If `Client.TableKey.getVersion() == io.pravega.client.tables.impl.TableKey.NOT_EXISTS` -> use `TableEntry.notExists`
3. Else use `TableKey.versioned`

It just occurred to me that ` io.pravega.client.tables.impl.TableKey.NOT_EXISTS` is null, which is incorrect (and has a TODO associated with it). We should set that to a different value. Let's say it has a version matching `Long.MIN_VALUE` so it matches the server-side.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java [hunk 17]
```diff
@@ -552,6 +571,144 @@ public void updateSegmentPolicy(UpdateSegmentPolicy updateSegmentPolicy) {
...
+        final String operation = "removeTableKeys";
+
+        if (!verifyToken(segment, removeTableKeys.getRequestId(), removeTableKeys.getDelegationToken(), operation)) {
+            return;
+        }
+
+        log.info(removeTableKeys.getRequestId(), "Removing table keys {}.", removeTableKeys);
+
+        List<TableKey> keys = removeTableKeys.getKeys().stream()
+                                                .map(k -> TableKey.versioned(new ByteArraySegment(k.getData()), k.getKeyVersion()))
```
Comment (andreipaduroiu) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: Same here as in `put`

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java (hunk 15 above)
Comment (andreipaduroiu) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: Yes, this has to be done.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java [hunk 18]
```diff
@@ -604,6 +761,12 @@ private Void handleException(long requestId, String segment, String operation, T
         } else if (u instanceof BadOffsetException) {
             BadOffsetException badOffset = (BadOffsetException) u;
             invokeSafely(connection::send, new SegmentIsTruncated(requestId, segment, badOffset.getExpectedOffset(), clientReplyStackTrace), failureHandler);
+        } else if (u instanceof TableSegmentNotEmptyException) {
+            log.warn(requestId, "Table segment '{}' is not empty to perform '{}'.", segment, operation);
+            invokeSafely(connection::send, new TableSegmentNotEmpty(requestId, segment, clientReplyStackTrace), failureHandler);
+        } else if (u instanceof ConditionalTableUpdateException) {
```
Comment (andreipaduroiu) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: I believe @shiveshr may want to distinguish between `BadKeyVersionException` and `KeyNotExistsException` (both of which derive from `ConditionalTableUpdateException`. I suggest we add special handlers for those too, above this one (and keep `ConditionalTableUpdateException`.

File: shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java [hunk 19]
```diff
@@ -1349,4 +1505,315 @@ public static WireCommand readFrom(ByteBufInputStream in, int length) throws IOE
...
+        }
+
+        public static TableKey readFrom(DataInput in, int length) throws IOException {
+            int payLoadSize = in.readInt();
+            int dataLength = in.readInt();
+            if (dataLength == 0) {
+                return TableKey.EMPTY;
+            }
+            if (length < payLoadSize) {
+                throw new InvalidMessageException("Was expecting length of atleast : " + payLoadSize + " but found: " + length);
```
Comment (andreipaduroiu) [inline shared/protocol/src/main/java/io/pravega/shared/protocol/netty/WireCommands.java]: `atleast` -> `at least`

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java (hunk 16 above)
Comment (shrids) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: My understanding was that the Table entry can have 3 possible versions.
a.  `TableKey.NO_VERSION `  (Long.MIN_VALUE)
b. `TableKey.NOT_EXISTS` ( -1L)
c. actual version. Any value >= 0L.
These are the 3 values which will always be sent inside a WireCommand and null is never passed as part of the WireCommand. Hence the `TableEntry.versioned(..)` api is used.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java (hunk 17 above)
Comment (shrids) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: same comment as in `put`.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java (hunk 18 above)
Comment (shrids) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: Created `WireCommands.TableKeyDoesNotExist` and `WireCommands.TableKeyBadVersion`.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java (hunk 15 above)
Comment (shrids) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: done.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java (hunk 16 above)
Comment (andreipaduroiu) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java]: Ah I see. Ok, then we should handle this when implementing the client-side contracts.
